`is_non_empty_f_string` has false positives

### Summary

[`is_non_empty_f_string`](https://github.com/astral-sh/ruff/blob/0.14.1/crates/ruff_python_ast/src/helpers.rs#L1350) has false positives for f-strings with certain interpolations. The interpolation `{f""!s}` is empty. Interpolations with format specs or containing comparison expressions have unknown emptiness. [Here is an example](https://play.ruff.rs/5612a49d-2c07-4852-9725-340777c1982a) with [`call-with-shell-equals-true` (S604)](https://docs.astral.sh/ruff/rules/call-with-shell-equals-true/):
```console
$ cat >is_non_empty_f_string_fp.py <<'# EOF'
print(bool(dict(shell=f"{f""!s}")["shell"]))

print(bool(dict(shell=f"{"x":.0}")["shell"]))

class C:
    def __init__(self, chars):
        self.chars = chars
    def __gt__(self, other):
        return "".join(sorted(c for c in self.chars if all(c > o for o in other.chars)))
print(bool(dict(shell=f"{C("abc") > C("xyz")}")["shell"]))
# EOF

$ python is_non_empty_f_string_fp.py
False
False
False

$ ruff --isolated check is_non_empty_f_string_fp.py --select S604 --output-format concise -q
is_non_empty_f_string_fp.py:1:12: S604 Function call with truthy `shell` parameter identified, security issue
is_non_empty_f_string_fp.py:3:12: S604 Function call with truthy `shell` parameter identified, security issue
is_non_empty_f_string_fp.py:10:12: S604 Function call with truthy `shell` parameter identified, security issue
```

### Version

ruff 0.14.1 (2bffef596 2025-10-16)

--- a/crates/ruff_linter/resources/test/fixtures/flake8_simplify/SIM222.py
+++ b/crates/ruff_linter/resources/test/fixtures/flake8_simplify/SIM222.py
@@ -228,3 +228,12 @@ tuple(None) or True  # OK
 tuple(...) or True  # OK
 tuple(lambda x: x) or True  # OK
 tuple(x for x in range(0)) or True  # OK
+
+# An f-string is guaranteed non-empty if any element is non-empty, even if others are indeterminate
+print(f"{x}abc" or "bar")  # SIM222
+print(f"abc{x}" or "bar")  # SIM222
+print(f"{x}" f"abc" or "bar")  # SIM222
+print(f"{x}{y}" or "bar")  # OK
+# An empty f-string part in an implicit concatenation guarantees nothing
+print(f"" f"{x}" or "bar")  # OK
+print(f"{f''}" or "bar")  # SIM222 (the f-string is provably empty, so `or` yields "bar")

--- a/crates/ruff_linter/resources/test/fixtures/flake8_simplify/SIM222.py
+++ b/crates/ruff_linter/resources/test/fixtures/flake8_simplify/SIM222.py
@@ -231,9 +231,9 @@ tuple(x for x in range(0)) or True  # OK
 
 # An f-string is guaranteed non-empty if any element is non-empty, even if others are indeterminate
 print(f"{x}abc" or "bar")  # SIM222
-print(f"abc{x}" or "bar")  # SIM222
-print(f"{x}" f"abc" or "bar")  # SIM222
-print(f"{x}{y}" or "bar")  # OK
 # An empty f-string part in an implicit concatenation guarantees nothing
 print(f"" f"{x}" or "bar")  # OK
-print(f"{f''}" or "bar")  # SIM222 (the f-string is provably empty, so `or` yields "bar")
+
+# Regression test for: https://github.com/astral-sh/ruff/issues/21048
+print(f"{'x':.0}" or "bar")  # OK
+print(f"{x > y}" or "bar")  # OK

--- a/crates/ruff_python_ast/src/helpers.rs
+++ b/crates/ruff_python_ast/src/helpers.rs
@@ -1569,7 +1569,6 @@ fn is_non_empty_f_string(expr: &ast::ExprFString) -> bool {
             Expr::ListComp(_) => true,
             Expr::SetComp(_) => true,
             Expr::DictComp(_) => true,
-            Expr::Compare(_) => true,
             Expr::NumberLiteral(_) => true,
             Expr::BooleanLiteral(_) => true,
             Expr::NoneLiteral(_) => true,
@@ -1586,6 +1585,8 @@ fn is_non_empty_f_string(expr: &ast::ExprFString) -> bool {
             Expr::BoolOp(ast::ExprBoolOp { .. }) => false,
             Expr::BinOp(ast::ExprBinOp { .. }) => false,
             Expr::UnaryOp(ast::ExprUnaryOp { .. }) => false,
+            // Rich comparison methods can return arbitrary objects.
+            Expr::Compare(_) => false,
             Expr::Generator(_) => false,
             Expr::Await(_) => false,
             Expr::Yield(_) => false,
@@ -1615,7 +1616,8 @@ fn is_non_empty_f_string(expr: &ast::ExprFString) -> bool {
             f_string.elements.iter().any(|element| match element {
                 InterpolatedStringElement::Literal(string_literal) => !string_literal.is_empty(),
                 InterpolatedStringElement::Interpolation(f_string) => {
-                    f_string.debug_text.is_some() || inner(&f_string.expression)
+                    f_string.debug_text.is_some()
+                        || (f_string.format_spec.is_none() && inner(&f_string.expression))
                 }
             })
         }

--- a/crates/ruff_python_ast/src/helpers.rs
+++ b/crates/ruff_python_ast/src/helpers.rs
@@ -1611,7 +1611,8 @@ fn is_non_empty_f_string(expr: &ast::ExprFString) -> bool {
     expr.value.iter().any(|part| match part {
         ast::FStringPart::Literal(string_literal) => !string_literal.is_empty(),
         ast::FStringPart::FString(f_string) => {
-            f_string.elements.iter().all(|element| match element {
+            // The part is a concatenation of elements, so it's guaranteed non-empty if any element is
+            f_string.elements.iter().any(|element| match element {
                 InterpolatedStringElement::Literal(string_literal) => !string_literal.is_empty(),
                 InterpolatedStringElement::Interpolation(f_string) => {
                     f_string.debug_text.is_some() || inner(&f_string.expression)

